File: python/2024/day_1.py
import os

def part_one(input: str) -> int:
    left, right = [], []

    original = [x for sublist in [[int(x) for x in line.split()] for line in input.strip().splitlines()] for x in sublist]
    left, right = original[::2], original[1::2]

    return sum(abs(x - y) for x,y in zip(sorted(left), sorted(right)))

def part_two(input: str) -> int:
    lines = input.strip().split('\n')

    lot = {}
    left = []

    for line in lines:
        (a,b) = list(filter(len, line.strip().split(" ")))

        left.append(int(a))

        if b in lot:
            lot[b] += 1
        else:
            lot[b] = 1

    sum = 0

    for value in left:
        if str(value) in lot:
            sum += value * lot[str(value)]

    return sum

dir = os.path.dirname(__file__);
input = open(os.path.join(dir, "day_1_input.txt")).read()

print("Day 1 | Part 1: ", part_one(input))
print("Day 1 | Part 2", part_two(input))

File: go/2024/day_1.go
package day_1

import (
	"math"
	"sort"
	"strconv"
	"strings"
)

func partOne(input string) int {
	lines := strings.Split(strings.TrimSpace(input), "\n")

	var sum float64 = 0

	left := []int{}
	right := []int{}

	for _, line := range lines {
		ele := strings.Fields(strings.TrimSpace(line))

		l, _ := strconv.Atoi(ele[0])
		r, _ := strconv.Atoi(ele[1])

		left = append(left, l)
		right = append(right, r)
	}

	sort.Ints(left)
	sort.Ints(right)

	for i, lValue := range left {
		rValue := right[i]

		sum += math.Abs(float64(lValue) - float64(rValue))
	}

	return int(sum)
}

func partTwo(input string) int {
	lot := make(map[int]int)
	lines := strings.Split(strings.TrimSpace(input), "\n")

	left := []int{}

	for _, line := range lines {
		ele := strings.Fields(strings.TrimSpace(line))

		l, _ := strconv.Atoi(ele[0])
		r, _ := strconv.Atoi(ele[1])

		_, exists := lot[r]

		if exists {
			lot[r] += 1
		} else {
			lot[r] = 1
		}

		left = append(left, l)
	}

	sum := 0

	for _, value := range left {
		count, exists := lot[value]

		if exists {
			sum += value * count
		}
	}

	return int(sum)
}
